### DISCUSSION TITLE
Under what circumstances would OT use CacheBasedVirtualField?

### DISCUSSION THREAD

**Discussion by 123liuziming** [open-telemetry/opentelemetry-java-instrumentation#9942] [Q&A] [answered]
Usually OT will use RuntimeVirtualField whenever possible. I want to know under what circumstances would OT use CacheBasedVirtualField?

**Comment by laurit** [comment]
It will be use when `VirtualField` is used in a library instrumentation.

**Comment by 123liuziming** [comment]
![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/32130965/7a820a2b-0dea-43d9-8ae3-314753dfff70)
This is the virtual field in rocketmq library instrumentation. It seems that it is a RuntimeVirtualField. Is there an example to use CacheBasedVirtualField? I 've not seen it yet.

**Reply by laurit** [reply, marked as answer]
Try using a library instrumentation without the javaagent.

**Reply by 123liuziming** [reply]
I agree, in this case, the constructor of VirtualFieldImplementationInstallerFactory will not be called.

**Comment by 123liuziming** [comment]
I think the CacheBasedVirtualField may be used for class that has already been loaded too early.
![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/32130965/981c665b-95ca-4af1-b909-40b53d58064b)
At that time the `RuntimeVirtualFieldSupplier.set` method is not executed.

**Comment by 123liuziming** [comment]
> Try using a library instrumentation without the javaagent.
